code/backend/batch/utilities/orchestrator/SemanticKernel.py

```diff
@@ -0,0 +1,137 @@
+from semantic_kernel.contents import ChatHistory
+
+from ..common.Answer import Answer
+from ..helpers.LLMHelper import LLMHelper
+from ..plugins.ChatPlugin import ChatPlugin
+from ..plugins.PostAnsweringPlugin import PostAnsweringPlugin
+from .OrchestratorBase import OrchestratorBase
+
+logger = logging.getLogger(__name__)
+
+
+class SemanticKernelOrchestrator(OrchestratorBase):
+    def __init__(self) -> None:
+        super().__init__()
+        self.kernel = Kernel()
+        self.llm_helper = LLMHelper()
+
+        # Add the Azure OpenAI service to the kernel
+        self.chat_service = self.llm_helper.get_sk_chat_completion_service("aoai")
+        self.kernel.add_service(self.chat_service)
+
+        self.kernel.add_plugin(
+            plugin=PostAnsweringPlugin(), plugin_name="PostAnswering"
+        )
+
+    async def orchestrate(
+        self, user_message: str, chat_history: list[dict], **kwargs: dict
+    ) -> list[dict]:
+        # Call Content Safety tool
+        if self.config.prompts.enable_content_safety:
+            if response := self.call_content_safety_input(user_message):
+                return response
+
+        system_message = """You help employees to navigate only private information sources.
+You must prioritize the function call over your general knowledge for any question by calling the search_documents function.
+Call the text_processing function when the user request an operation on the current context, such as translate, summarize, or paraphrase. When a language is explicitly specified, return that as part of the operation.
+When directly replying to the user, always reply in the language the user is speaking."""
+
+        self.kernel.add_plugin(
+            plugin=ChatPlugin(question=user_message, chat_history=chat_history),
+            plugin_name="Chat",
+        )
+
+        settings = self.llm_helper.get_sk_service_settings(self.chat_service)
+        settings.tools = get_tool_call_object(
+            kernel=self.kernel, filter={"include_plugin": ["Chat"]}
+        )
+
+        orchestrate_function = self.kernel.add_function(
+            plugin_name="Main",
+            function_name="orchestrate",
+            prompt="{{$chat_history}}{{$user_message}}",
+            prompt_execution_settings=settings,
+        )
+
+        history = ChatHistory(system_message=system_message)
+
+        for message in chat_history.copy():
+            history.add_message(message)
+
+        result = (
+            await self.kernel.invoke(
+                function=orchestrate_function,
+                chat_history=history,
+                user_message=user_message,
+            )
+        ).value[0]
+
+        self.log_tokens(
+            prompt_tokens=result.metadata["usage"].prompt_tokens,
+            completion_tokens=result.metadata["usage"].completion_tokens,
+        )
+
+        if result.tool_calls:
+            logger.info("Function call detected")
```

Comment:
Do you think it's worth having some reference of "sk" in here?

Reply:
Updated